src/main/scala/edu/cmu/cs/obsidian/codegen/CodeGenYul.scala

```diff
@@ -173,32 +192,50 @@ object CodeGenYul extends CodeGenerator {
                 new_name, // TODO rename transaction name (by adding prefix/suffix) iev: this seems to be done already
                 constructor.args.map(v => TypedName(v.varName, mapObsTypeToABI(v.typIn.toString))),
                 Seq(), //todo/iev: why is this always empty?
-                Block(constructor.body.flatMap((s: Statement) => translateStatement(s))))) //todo iev temp vars: likely a hack to work around something wrong
+                Block(constructor.body.flatMap((s: Statement) => translateStatement(s, None))))) //todo iev flatmap may be a bug to hide something wrong; None means that constructors don't return. is that true?
     }
 
     def translateTransaction(transaction: Transaction): Seq[YulStatement] = {
-        val ret: Seq[TypedName] =
-            if (transaction.retType.isEmpty) {
-                Seq()
-            } else { // TODO hard code return variable name now, need a special naming convention to avoid collisions
-                Seq(TypedName("retValTempName", mapObsTypeToABI(transaction.retType.get.toString)))
+        var id: Option[String] = None
+        val ret: Seq[TypedName] = {
+            transaction.retType match {
+                case Some(t) =>
+                    id = Some(nextRet())
+                    Seq(TypedName(id.get, mapObsTypeToABI(t.toString)))
+                case None => Seq()
             }
+        }
+
+        // update the mapping for returns
+        fnNamesToReturnArity = fnNamesToReturnArity.+((transaction.name, ret.length))
 
         Seq(FunctionDefinition(
             transaction.name, // TODO rename transaction name (by adding prefix/suffix)
             transaction.args.map(v => TypedName(v.varName, mapObsTypeToABI(v.typIn.toString))),
             ret,
-            Block(transaction.body.flatMap((s: Statement) => translateStatement(s))))) //todo iev temp vars: likely a hack to work around something wrong
+            Block(transaction.body.flatMap((s: Statement) => translateStatement(s, id))))) //todo iev temp vars: likely a hack to work around something wrong
     }
 
-    def translateStatement(s: Statement): Seq[YulStatement] = {
+    /**
+      * translate an obsidian statement into the corresponding sequence of yul statements
+      *
+      * @param s      the statement to be translated
+      * @param retVar the name of the variable to use for returning for the current scope, if there is one
+      * @return
+      */
+    def translateStatement(s: Statement, retVar: Option[String]): Seq[YulStatement] = {
```

Comment:
Be consistent about the order. translateExpr has the variable as the first arg, and it's a different type. Or is there a reason these are different?